shared/params.go

```diff
@@ -58,17 +59,100 @@ func ParseSHAParamFull(r *http.Request) (runSHA string, err error) {
 	return runSHA, err
 }
 
+// ParseProductAtRevision parses a test-run spec into a ProductAtRevision struct.
+func ParseProductAtRevision(spec string) (productAtRevision ProductAtRevision, err error) {
+	pieces := strings.Split(spec, "@")
+	if len(pieces) > 2 {
+		return productAtRevision, errors.New("invalid product@revision spec: " + spec)
+	}
+	if productAtRevision.Product, err = ParseProduct(pieces[0]); err != nil {
+		return productAtRevision, err
+	}
+	if len(pieces) < 2 {
+		// No @ is assumed to be the product only.
+		productAtRevision.Revision = "latest"
+	} else {
+		productAtRevision.Revision = pieces[1]
+	}
+	return productAtRevision, nil
+}
+
+// ParseProduct parses the `browser-version-os-version` input as a Product struct.
+func ParseProduct(product string) (result Product, err error) {
+	pieces := strings.Split(product, "-")
+	if len(pieces) > 4 {
+		return result, fmt.Errorf("invalid product: %s", product)
+	}
+	result = Product{
+		BrowserName: pieces[0],
+	}
+	if !IsBrowserName(result.BrowserName) {
+		return result, fmt.Errorf("invalid browser name: %s", result.BrowserName)
+	}
+	if len(pieces) > 1 {
+		if _, err := ParseVersion(pieces[1]); err != nil {
+			return result, fmt.Errorf("invalid browser version: %s", pieces[1])
+		}
+		result.BrowserVersion = pieces[1]
+	}
+	if len(pieces) > 2 {
+		result.OSName = pieces[2]
+	}
+	if len(pieces) > 3 {
+		if _, err := ParseVersion(pieces[3]); err != nil {
+			return result, fmt.Errorf("invalid OS version: %s", pieces[3])
+		}
+		result.OSVersion = pieces[3]
+	}
+	return result, nil
+}
+
+// ParseVersion parses the given version as a semantically versioned string.
+func ParseVersion(version string) (result *Version, err error) {
+	pieces := strings.Split(version, ".")
+	for i, piece := range pieces {
+		if _, err := strconv.ParseInt(piece, 10, 0); i > 3 || err != nil {
+			return nil, fmt.Errorf("Invalid version: %s", version)
+		}
+	}
+	result = &Version{
+		Major: pieces[0],
+	}
+	if len(pieces) > 1 {
+		result.Minor = pieces[1]
+	}
+	if len(pieces) > 2 {
+		result.Revision = pieces[2]
+	}
+	return result, nil
+}
+
+// ParseProductParam parses and validates the 'product' param for the request.
+func ParseProductParam(r *http.Request) (product *Product, err error) {
+	productParam := r.URL.Query().Get("product")
+	if "" == productParam {
+		return nil, nil
+	}
+	parsed, err := ParseProduct(productParam)
```

Comment:
We treat the empty input as non-erroring above, so a pointer helps us spot `nil` return more easily.

This is an example of Golang being annoyingly verbose.

Reply:
I mean that if  `ParseProduct` itself is made to return a pointer, then this would be a bit simpler and a not-null-but-still-bogus return value wouldn't exist. Doesn't really matter much though.